parselogical/parselogical.go

```diff
@@ -0,0 +1,289 @@
+				}
+				pr.Operation = message[state.TokenStart:i]
+				if pr.Operation == "TRUNCATE" {
+					state.Current = parseStateOperationTruncate
+					break outer
+				}
+
+				state.TokenStart = i + 2
+				state.Current = parseStateColumnName
+
+				if preludeOnly {
+					break outer
+				}
+			}
+		case parseStateColumnName:
+			if chr == '[' {
+				state.CurColumnName = message[state.TokenStart:i]
+				state.TokenStart = i + 1
+				state.Current = parseStateColumnType
+			} else if chr == ':' {
+				if message[state.TokenStart:i] == "old-key" {
+					state.OldKey = true
+				} else if message[state.TokenStart:i] == "new-tuple" {
+					state.OldKey = false
+				}
+				state.TokenStart = i + 2
+			} else if chr == '(' && message[state.TokenStart:] == "(no-tuple-data)" {
+				pr.NoTupleData = true
+				state.Current = parseStateEnd
+			} else if chr == '"' {
+				state.Prev = state.Current
+				state.Current = parseStateEscapedIdentifier
+			}
+		case parseStateColumnType:
+			if chr == ']' {
+				if chrNext != ':' {
+					return errors.Errorf("invalid character '%s' at %d", []byte{chrNext}, i+1)
+				}
+				state.CurColumnType = message[state.TokenStart:i]
+				state.TokenStart = i + 2
+				state.Current = parseStateColumnValue
+			} else if chr == '"' {
+				state.Prev = state.Current
+				state.Current = parseStateEscapedIdentifier
+			} else if chr == '[' {
+				state.Prev = state.Current
+				state.Current = parseStateOpenSquareBracket
+			}
+		case parseStateColumnValue:
+			if chr == '\000' || chr == ' ' {
+				quoted := state.Prev == parseStateColumnQuotedValue
+				startStr := state.TokenStart
+				endStr := i
+
+				if quoted {
+					startStr++
+					endStr--
+				}
+
+				unescapedValue := strings.Replace(message[startStr:endStr], "''", "'", -1) // the value is escape-quoted to us
+
+				cv := ColumnValue{Value: unescapedValue, Quoted: quoted, Type: state.CurColumnType}
+
+				if state.OldKey {
+					pr.OldColumns[state.CurColumnName] = cv
+				} else {
+					pr.Columns[state.CurColumnName] = cv
+				}
+			}
+
+			if chr == '\000' {
+				state.Current = parseStateEnd
+			} else if chr == ' ' {
+				state.TokenStart = i + 1
+				state.Prev = state.Current
+				state.Current = parseStateColumnName
+			} else if chr == '\'' {
+				state.Prev = state.Current
+				state.Current = parseStateColumnQuotedValue
+			}
+		case parseStateOpenSquareBracket:
+			if chr == ']' {
+				state.Current = state.Prev
+				state.Prev = parseStateNull
+			}
+		case parseStateEscapedIdentifier:
+			if chr == '"' {
+				if chrNext == '"' {
+					i++
+				} else {
+					state.Current = state.Prev
+					state.Prev = parseStateNull
+				}
+			}
+		case parseStateColumnQuotedValue:
+			if chr == '\'' {
+				if chrNext == '\'' {
+					i++
+				} else {
+					prev := state.Prev
+					state.Prev = state.Current
+					state.Current = prev
+				}
+			}
+		}
+	}
+
+	if state.Current == parseStateOperationTruncate {
```

Comment:
it looks like the main change to the parser is to allow for messages with 'TRUNCATE' operations to be passed downstream to the rest of the pg-bifrost handlers. Presumably you've been able to run & test this end-to-end locally?

Reply:
oh nice, i see the additional itests as well.

Reply:
Yeah I added an integration test.